vulkan/vulkan_swapchain.cc

```diff
@@ -0,0 +1,511 @@
+      .compositeAlpha = VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR,
+      .presentMode = present_mode,
+      .clipped = VK_FALSE,
+      .oldSwapchain = old_swapchain_handle,
+  };
+
+  VkSwapchainKHR swapchain = VK_NULL_HANDLE;
+
+  if (VK_CALL_LOG_ERROR(vk.CreateSwapchainKHR(
+          device_.Handle(), &create_info, nullptr, &swapchain)) != VK_SUCCESS) {
+    return;
+  }
+
+  swapchain_ = {swapchain, [this](VkSwapchainKHR swapchain) {
+                  FTL_ALLOW_UNUSED_LOCAL(device_.WaitIdle());
+                  vk.DestroySwapchainKHR(device_.Handle(), swapchain, nullptr);
+                }};
+
+  if (!CreateSwapchainImages(skia_context)) {
+    return;
+  }
+
+  valid_ = true;
+}
+
+VulkanSwapchain::~VulkanSwapchain() = default;
+
+bool VulkanSwapchain::IsValid() const {
+  return valid_;
+}
+
+std::vector<VkImage> VulkanSwapchain::GetImages() const {
+  uint32_t count = 0;
+  if (VK_CALL_LOG_ERROR(vk.GetSwapchainImagesKHR(
+          device_.Handle(), swapchain_, &count, nullptr)) != VK_SUCCESS) {
+    return {};
+  }
+
+  if (count == 0) {
+    return {};
+  }
+
+  std::vector<VkImage> images;
+
+  images.resize(count);
+
+  if (VK_CALL_LOG_ERROR(vk.GetSwapchainImagesKHR(
+          device_.Handle(), swapchain_, &count, images.data())) != VK_SUCCESS) {
+    return {};
+  }
+
+  return images;
+}
+
+SkISize VulkanSwapchain::GetSize() const {
+  VkExtent2D extents = capabilities_.currentExtent;
+
+  if (extents.width < capabilities_.minImageExtent.width) {
+    extents.width = capabilities_.minImageExtent.width;
+  } else if (extents.width > capabilities_.maxImageExtent.width) {
+    extents.width = capabilities_.maxImageExtent.width;
+  }
+
+  if (extents.height < capabilities_.minImageExtent.height) {
+    extents.height = capabilities_.minImageExtent.height;
+  } else if (extents.height > capabilities_.maxImageExtent.height) {
+    extents.height = capabilities_.maxImageExtent.height;
+  }
+
+  return SkISize::Make(extents.width, extents.height);
+}
+
+sk_sp<SkSurface> VulkanSwapchain::CreateSkiaSurface(GrContext* gr_context,
+                                                    VkImage image,
+                                                    const SkISize& size) const {
+  if (gr_context == nullptr) {
+    return nullptr;
+  }
+
+  GrPixelConfig pixel_config = kUnknown_GrPixelConfig;
+
+  if (!GrVkFormatToPixelConfig(surface_format_.format, &pixel_config)) {
+    // Vulkan format unsupported by Skia.
+    return nullptr;
+  }
+
+  const GrVkImageInfo image_info = {
+      .fImage = image,
+      .fAlloc = {VK_NULL_HANDLE, 0, 0, 0},
+      .fImageTiling = VK_IMAGE_TILING_OPTIMAL,
+      .fImageLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+      .fFormat = surface_format_.format,
+      .fLevelCount = 1,
+  };
+
+  GrBackendRenderTargetDesc desc;
+  desc.fWidth = size.fWidth;
+  desc.fHeight = size.fHeight;
+  desc.fConfig = pixel_config;
+  desc.fOrigin = kTopLeft_GrSurfaceOrigin;
+  desc.fSampleCnt = 0;
```

Comment:
Skia would be happier with both `sampleCnt` and `stencilBits`. I will figure out the best flags and address in a later patch.